### PR TITLE
Refactor `rtc_cntl` module and add `get_reset_reason` function

### PR DESCRIPTION
**Pull request by jessebraham**
I realize this is a bit of an odd PR on its own, but I am building towards some other more interesting work. This was pretty self-contained so I figured I'd just PR it now.

The `rtc_cntl` module is going to grow quite a bit, so I've pulled it into its own module directory. The `rtc` module followed it into this directory as well.

The `get_reset_reason` function is pretty simple and basically just maps a ROM function's return value to an enum, `SocResetReason`. There were a few variants which had conflicting values, so I have noted these in comments and, IMO, provided a reasonable workaround, at least for the time being.

### CODE REVIEW

**Review by bjoernQ — APPROVED** [review]
LGTM

File: esp-hal-common/src/rom.rs [hunk 1]
```diff
@@ -127,3 +127,21 @@ macro_rules! regi2c_write_mask {
...
+    #[cfg(esp32c2)]
+    const RTC_GET_RESET_REASON: u32 = 0x4000_0018;
+    #[cfg(esp32c3)]
+    const RTC_GET_RESET_REASON: u32 = 0x4000_0018;
+    #[cfg(esp32s2)]
+    const RTC_GET_RESET_REASON: u32 = 0x4000_ff58;
+    #[cfg(esp32s3)]
+    const RTC_GET_RESET_REASON: u32 = 0x4000_057c;
+
+    let get_reason: fn(u32) -> u32 = core::mem::transmute(RTC_GET_RESET_REASON as usize);
```
Comment (MabezDev) [inline esp-hal-common/src/rom.rs:144]: This comment is orthogonal to the PR, I'm just curious. I believe we already do it this way in the clocks section & esp-wifi, but did we have a reason for transmuting addresses vs using `PROVIDE()` in each chips linker script?

**Review by MabezDev — APPROVED** [review]
LGTM!

File: esp-hal-common/src/rom.rs (hunk 1 above)
Comment (bjoernQ) [reply esp-hal-common/src/rom.rs:144]: I think in the beginning there was just like one function and adding something like `rom_functions.x` for each chip was a bit inconvenient - but the more rom functions we use that might be the nicer approach

Comment (jessebraham) [reply esp-hal-common/src/rom.rs:144]: Yeah I don't think there's any real reason to continue doing this, I was just copy-pasting as usual 😁  In a subsequent PR we can change this.
